### DISCUSSION TITLE
How to access uploadSequenceToken variable from response(DescribeLogStreamResponse) object

### DISCUSSION THREAD

**Discussion by RachaelSMathew** [soto-project/soto#574] [Q&A]
When I call describeLogStreams I am getting a nil value for the uploadSequenceToken. 

So, I am trying to catch the CloudWatchErrorType.invalidNextToken error so that I can get the uploadSequenceToken value. To produce the CloudWatchErrorType.invalidNextToken error I need to access the uploadSequenceToken from the response or DescribeLogStreamResponse object. 

<img width="869" alt="Screen Shot 2022-01-10 at 8 56 19 PM" src="https://user-images.githubusercontent.com/30049533/148867826-65ccf981-5bf6-4faa-999c-43a9e0b7af80.png">

**Comment by adam-fowler** [comment]
Hi @RachaelSMathew 

There are a number of issues with your code above. First any error thrown in the `whenComplete` closure will not be caught by the `catch` block. You need to chain your EventLoopFutures to catch errors eg

```swift
logClient.describeLogStreams(describeLogStreamsRequest)
    .flatMap { response in
        logClient.putEventStreams(...)
    }
    .flatMapErrorThrowing { error in
        // deal with error
    }
```
Reading the SwiftNIO docs on [EventLoopFuture](https://apple.github.io/swift-nio/docs/current/NIOCore/Classes/EventLoopFuture.html) may help.

Secondly you are trying to catch an error from the `CloudWatch` module related to the pagination tokens when calling a function in the `CloudWatchLogs` module. Are you wanting to catch `CloudWatchLogs.invalidSequenceTokenException`? 

Once you have a `CloudWatchLogs.invalidSequenceTokenException` error you can access the expectedSequenceToken via 
```swift
error.context.additionalFields["expectedSequenceToken"]
```

**Reply by adam-fowler** [reply]
Have you thought about using the new async/await APIs as they are easier to use.

```swift
let logs = try await logClient.describeLogStreams(describeLogStreamsRequest)
do {
    try await logClient.putEventStreams(...)
} catch let error as CloudWatchLogsErrorType where error == . expectedSequenceToken {
    print(error.context.additionalFields["expectedSequenceToken"])
}
```
